src/libs/antares/study/parallel-years.cpp

```diff
@@ -0,0 +1,431 @@
+    {
+        unsigned int indexSpace = 999999;
+        bool performCalculations = true;
+        
+        if(p.userPlaylist)
+            performCalculations = p.yearsFilter[y];
+        
+        bool refreshing = isRefreshNeededForCurrentYear(y);
+        buildNewSet = buildNewSet || refreshing;
+
+        if (buildNewSet)
+        {
+            setOfParallelYears setToCreate;
+            setsOfParallelYears.push_back(setToCreate);
+            set = &(setsOfParallelYears.back());
+
+            // Initializations
+            set->nbPerformedYears = 0;
+            set->nbYears = 0;
+            set->regenerateTS = false;
+            set->yearForTSgeneration = 999999;
+
+            // In case we have to regenerate times series before run the current set of parallel
+            // years
+            if (refreshing)
+            {
+                set->regenerateTS = true;
+                set->yearForTSgeneration = y;
+            }
+        }
+
+        set->yearsIndices.push_back(y); 
+        set->nbYears++;
+        set->yearFailed[y] = true;
+        set->isFirstPerformedYearOfASet[y] = false;
+
+        if (performCalculations)
+        {
+            set->setsSizes++;
+            // Another year performed
+            ++nbYearsReallyPerformed;
+
+            // Number of actually performed years in the current set (up to now).
+            set->nbPerformedYears++;
+            // Index of the MC year's space (useful if this year is actually run)
+            indexSpace = set->nbPerformedYears - 1;
+
+            set->isYearPerformed[y] = true;
+            set->performedYearToSpace[y] = indexSpace;
+            set->spaceToPerformedYear[indexSpace] = y;
+
+            if (!foundFirstPerformedYearOfCurrentSet)
+            {
+                set->isFirstPerformedYearOfASet[y] = true;
+                foundFirstPerformedYearOfCurrentSet = true;
+            }
+        }
+        else
+        {
+            set->isYearPerformed[y] = false;
+        }
+
+        // Do we build a new set at next iteration (for years to be executed or not) ?
+
+        // In case the study is run in the draft mode, only 1 core is allowed
+        if (p.mode == Antares::Data::stdmAdequacyDraft){
+            forcedNbOfParallelYears_ = 1;
+        }
+
+        // In case parallel mode was not chosen, only 1 core is allowed
+        if (!enableParallel_ && !forceParallel_){
+            forcedNbOfParallelYears_ = 1;
+        }
+
+        if (indexSpace == forcedNbOfParallelYears_ - 1 || y == p.nbYears - 1)
+        {
+            buildNewSet = true;
+            foundFirstPerformedYearOfCurrentSet = false;
+            if (set->nbPerformedYears > forcedNbOfParallelYears_)
+                forcedNbOfParallelYears_ = set->nbPerformedYears;
+        }
+        else
+            buildNewSet = false;
+
+     // End of loop over years
+
+    }
+
+}
+
+
+bool SetsOfParallelYearCalculator::allSetsParallelYearsHaveSameSize()
```

Comment:
About **allSetsParallelYearsHaveSameSize()** : 
This function answer the question : does all sets contains the same number of performed years ?
For now It is only called in case the study enables the hydro hot start. 
And that is reflected in the body of this function.
Well it should not : the function should stick to its name and answer the simple question without knowing if it's called in the context of hydro hot start or not.
So the body of this function should be : 
```C++
bool SetsOfParallelYearCalculator::allSetsParallelYearsHaveSameSize()
{
    uint currentSetSize = (uint)setsOfParallelYears[0].setsSizes;
    return all_of(setsOfParallelYears.begin(), setsOfParallelYears.end(),
                    [currentSetSize](const setOfParallelYears& v){ return v.setsSizes == currentSetSize; } );
}
```
and that's it.
Another benefit of this simplification is that  `p.initialReservoirLevels.iniLevels` is not required anymore in class **SetsOfParallelYearCalculator**, so we loose a dependency on a study parameter.